### PR TITLE
[main] [feat] auto stake hlp

### PR DESCRIPTION
**Pull request by john-xina-p88**

### CODE REVIEW

File: script/ts/configs/LiquidityHandler/set-hlp-staking.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+import { DeployFunction } from "hardhat-deploy/types";
+import { ethers } from "hardhat";
```
Comment (shutsuwei) [inline script/ts/configs/LiquidityHandler/set-hlp-staking.ts:3]: `import { ethers } form "ethers"` instead

File: script/ts/configs/LiquidityHandler/set-hlp-staking.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import { EvmPriceServiceConnection } from "@pythnetwork/pyth-evm-js";
```
Comment (shutsuwei) [inline script/ts/configs/LiquidityHandler/set-hlp-staking.ts:4]: Remove unused deps

File: script/ts/configs/LiquidityHandler/set-hlp-staking.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import {
+  IPyth__factory,
+  LimitTradeHandler__factory,
+  LiquidityHandler__factory,
+  MockPyth__factory,
+  PythAdapter__factory,
+} from "../../../../typechain";
+import { getConfig } from "../../utils/config";
```
Comment (shutsuwei) [inline script/ts/configs/LiquidityHandler/set-hlp-staking.ts:12]: use loadConfig(chainId) instead

File: script/ts/configs/LiquidityHandler/set-hlp-staking.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+const config = getConfig();
+const BigNumber = ethers.BigNumber;
+const parseUnits = ethers.utils.parseUnits;
+
+async function main() {
+  const deployer = (await ethers.getSigners())[0];
```
Comment (shutsuwei) [inline script/ts/configs/LiquidityHandler/set-hlp-staking.ts:19]: use `signers.deployer(chainId)` instead

File: script/ts/configs/LiquidityHandler/set-hlp-staking.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+  console.log("> LiquidityHandler: Set HLP Staking...");
+  const handler = LiquidityHandler__factory.connect(config.handlers.liquidity, deployer);
+  await (await handler.setHlpStaking(config.staking.hlp)).wait();
+  console.log("> LiquidityHandler: Set HLP Staking success!");
+}
+main().catch((error) => {
```
Comment (shutsuwei) [inline script/ts/configs/LiquidityHandler/set-hlp-staking.ts:26]: gracefully exit by adding
`main().then(() => process.exit(0)).catch...`

**Review by jinlong9922 — APPROVED** [review]

**Review by yingyang888 — APPROVED** [review]

File: src/handlers/LiquidityHandler.sol [hunk 6]
```diff
@@ -187,7 +218,7 @@ contract LiquidityHandler is OwnableUpgradeable, ReentrancyGuardUpgradeable, ILi
         status: LiquidityOrderStatus.PENDING
       })
     );
-
+    isHyper[msg.sender][_orderId] = _isHyper;
```
Comment (shutsuwei) [inline src/handlers/LiquidityHandler.sol:221]: isSurge

File: src/handlers/LiquidityHandler.sol [hunk 7]
```diff
@@ -153,24 +159,49 @@ contract LiquidityHandler is OwnableUpgradeable, ReentrancyGuardUpgradeable, ILi
...
+    return _createAddLiquidityOrder(_tokenIn, _amountIn, _minOut, _executionFee, _shouldWrap, _isHyper);
+  }
+
+  function _createAddLiquidityOrder(
+    address _tokenIn,
+    uint256 _amountIn,
+    uint256 _minOut,
+    uint256 _executionFee,
+    bool _shouldWrap,
+    bool _isHyper
```
Comment (shutsuwei) [inline src/handlers/LiquidityHandler.sol:189]: isSurge

File: src/handlers/LiquidityHandler.sol [hunk 8]
```diff
@@ -153,24 +159,49 @@ contract LiquidityHandler is OwnableUpgradeable, ReentrancyGuardUpgradeable, ILi
...
+  /// @param _executionFee The execution fee of order
+  /// @param _shouldWrap in case of sending native token
+  /// @param _isHyper true will participate this liquidity into HLP Surge
+  function createAddLiquidityOrder(
+    address _tokenIn,
+    uint256 _amountIn,
+    uint256 _minOut,
+    uint256 _executionFee,
+    bool _shouldWrap,
+    bool _isHyper
```
Comment (shutsuwei) [inline src/handlers/LiquidityHandler.sol:178]: isSurge

File: src/handlers/LiquidityHandler.sol [hunk 9]
```diff
@@ -364,13 +395,37 @@ contract LiquidityHandler is OwnableUpgradeable, ReentrancyGuardUpgradeable, ILi
...
+        );
+        if (isHyper[_order.account][_order.orderId]) {
+          IHyperStaking(hlpStaking).depositHyper(_order.account, _amountOut);
+        } else {
+          IHyperStaking(hlpStaking).deposit(_order.account, _amountOut);
+        }
+      } else {
+        // If no HLPStaking
+        IERC20Upgradeable(_order.token).safeTransfer(LiquidityService(liquidityService).vaultStorage(), _order.amount);
+        _amountOut = LiquidityService(liquidityService).addLiquidity(
```
Comment (shutsuwei) [inline src/handlers/LiquidityHandler.sol:422]: you can combine addLiquidity part to 1 block and have ternary to decide whether which address to receive the HLP

File: src/handlers/LiquidityHandler.sol [hunk 10]
```diff
@@ -364,13 +395,37 @@ contract LiquidityHandler is OwnableUpgradeable, ReentrancyGuardUpgradeable, ILi
...
+          address(this)
+        );
+
+        // Auto stake into HLPStaking
+        IERC20Upgradeable(ConfigStorage(LiquidityService(liquidityService).configStorage()).hlp()).safeApprove(
+          address(hlpStaking),
+          _amountOut
+        );
+        if (isHyper[_order.account][_order.orderId]) {
+          IHyperStaking(hlpStaking).depositHyper(_order.account, _amountOut);
```
Comment (shutsuwei) [inline src/handlers/LiquidityHandler.sol:415]: depositSurge

File: src/handlers/LiquidityHandler.sol [hunk 11]
```diff
@@ -364,13 +395,37 @@ contract LiquidityHandler is OwnableUpgradeable, ReentrancyGuardUpgradeable, ILi
...
+        _amountOut = LiquidityService(liquidityService).addLiquidity(
+          _order.account,
+          _order.token,
+          _order.amount,
+          _order.minOut,
+          address(this)
+        );
+
+        // Auto stake into HLPStaking
+        IERC20Upgradeable(ConfigStorage(LiquidityService(liquidityService).configStorage()).hlp()).safeApprove(
```
Comment (shutsuwei) [inline src/handlers/LiquidityHandler.sol:410]: Can move approval to maxUint256 when set HLPStaking

**Review by shutsuwei — CHANGES_REQUESTED** [review]

File: script/ts/configs/LiquidityHandler/set-hlp-staking.ts (hunk 1 above)
Comment (john-xina-p88) [reply script/ts/configs/LiquidityHandler/set-hlp-staking.ts:3]: There is no `getSigners()` in that ethers.

File: src/handlers/LiquidityHandler.sol (hunk 9 above)
Comment (john-xina-p88) [reply src/handlers/LiquidityHandler.sol:422]: I'm gonna keep it as is. Changing this, make old test failed.

File: src/handlers/LiquidityHandler.sol [hunk 12]
```diff
@@ -93,6 +96,9 @@ contract LiquidityHandler is OwnableUpgradeable, ReentrancyGuardUpgradeable, ILi
   mapping(address => bool) public orderExecutors; // address -> whitelist executors
   mapping(address => LiquidityOrder[]) public accountExecutedLiquidityOrders; // account -> executed orders
 
+  IHyperStaking public hlpStaking;
```
Comment (shutsuwei) [inline src/handlers/LiquidityHandler.sol:99]: `IHlpStaking`

File: src/staking/interfaces/IHyperStaking.sol [hunk 13]
```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: BUSL-1.1
+// This code is made available under the terms and conditions of the Business Source License 1.1 (BUSL-1.1).
+// The act of publishing this code is driven by the aim to promote transparency and facilitate its utilization for educational purposes.
+
+pragma solidity 0.8.18;
+
+interface IHyperStaking {
```
Comment (shutsuwei) [inline src/staking/interfaces/IHyperStaking.sol:7]: `Hyper` -> `Surge`

File: src/staking/interfaces/IHyperStaking.sol [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+  struct TierConfig {
+    uint256 maxCap;
+    uint256 multiplier;
+  }
+
+  function startHyperEventDepositTimestamp() external view returns (uint256);
```
Comment (shutsuwei) [inline src/staking/interfaces/IHyperStaking.sol:13]: `startSurge`

**Review by shutsuwei — CHANGES_REQUESTED** [review]

**Review by shutsuwei — APPROVED** [review]

**Pull request merged**